pkg/controller/volume/expand/expand_controller.go

```diff
@@ -302,19 +305,31 @@ func (expc *expandController) expand(pvc *v1.PersistentVolumeClaim, pv *v1.Persi
 		return util.DeleteAnnPreResizeCapacity(pv, expc.GetKubeClient())
 	}
 
-	pvc, err := util.MarkResizeInProgressWithResizer(pvc, resizerName, expc.kubeClient)
-	if err != nil {
-		klog.V(5).Infof("Error setting PVC %s in progress with error : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), err)
-		return err
-	}
+	var generatedOptions volumetypes.GeneratedOperations
+	var err error
 
-	generatedOperations, err := expc.operationGenerator.GenerateExpandVolumeFunc(pvc, pv)
-	if err != nil {
-		klog.Errorf("Error starting ExpandVolume for pvc %s with %v", util.GetPersistentVolumeClaimQualifiedName(pvc), err)
-		return err
+	if utilfeature.DefaultFeatureGate.Enabled(features.RecoverVolumeExpansionFailure) {
+		generatedOptions, err = expc.operationGenerator.GenerateExpandAndRecoverVolumeFunc(pvc, pv, resizerName)
+		if err != nil {
+			klog.Errorf("Error starting ExpandVolume for pvc %s with %v", util.GetPersistentVolumeClaimQualifiedName(pvc), err)
+			return err
+		}
+	} else {
+		pvc, err := util.MarkResizeInProgressWithResizer(pvc, resizerName, expc.kubeClient)
```

Comment:
this Mark is only needed if feature is not enabled?

Reply:
The mark is needed in both the cases but in case this new feature is enabled, we are doing everything within the generator function. i.e - marking pvc for resize in progress and set new `ResizeStatus`. The reason I moved everything in one place is because we want all changes to PVC submitted in single API request .